### PR TITLE
DRILL-7156: Support empty Parquet files creation

### PR DESCRIPTION
**Pull request by oleg-zinovev**
PR for Drill empty parquet  files read and write support.

Known limitations:
1) Not working for hive parquet for now
2) Ignores all schemas except last while writing empty parquet file
3) Not support empty schemas (e.g. create table `..` as select * from `empty.json, e.g. {}`)

Short changes description:
1) Parquet footer metadata added
2) Parquet writer checks that at least 1 row has been written. If not - creates a empty parquet file with footer.
3) EmptyParquetRowGroupScan and EmptyParquetScanBatchCreator added

Questions:
1) TestParquetWriterEmptyFiles#testMultipleWriters now creates several empty files, but not fails, since reading of empty parquet is supported. Should I rewrite comment or remove the test?

### CODE REVIEW

**Comment by oleg-zinovev** [comment]
@arina-ielchiieva, could you please review?

**Comment by arina-ielchiieva** [comment]
@oleg-zinovev thanks for making the changes, though the situation is little bit awkward, since I was working on similar changes and did not know you intend to do them as well (https://issues.apache.org/jira/browse/DRILL-4517). Though I was working on reading empty parquet files but not writing them. I suggest you separate out writing empty parquet files into separate PR as for reading it might be better if my changes will be used instead: first you change metadata cache files and this would affect backward compatibility as well as will have to store more information than needed, secondly your changes does not seem to optimize reading complex types. What do you think?

**Comment by oleg-zinovev** [comment]
Ok, I'll try to rewrite the commit within a week.

**Comment by arina-ielchiieva** [comment]
@oleg-zinovev thanks, I have assigned https://issues.apache.org/jira/browse/DRILL-7156 to you.

**Comment by arina-ielchiieva** [comment]
@oleg-zinovev support empty parquet files reading is already merged into master (https://github.com/apache/drill/commit/4f4e1af53c9abccd1996f3b6841731e68768b48e). Do you plan on working on adding support for writing empty parquet files? We plan to include it in next Drill release (end of August / beginning of September). If yes, please factor out writing empty parquet and update the PR.

**Comment by oleg-zinovev** [comment]
> @oleg-zinovev support empty parquet files reading is already merged into master ([4f4e1af](https://github.com/apache/drill/commit/4f4e1af53c9abccd1996f3b6841731e68768b48e)). Do you plan on working on adding support for writing empty parquet files? We plan to include it in next Drill release (end of August / beginning of September). If yes, please factor out writing empty parquet and update the PR.

Yes, I think I’ll do it today.

**Comment by arina-ielchiieva** [comment]
Regarding this comment

> Questions:
TestParquetWriterEmptyFiles#testMultipleWriters now creates several empty files, but not fails, since reading of empty parquet is supported. Should I rewrite comment or remove the test?

I guess you can remove these tests and add new tests into `org.apache.drill.exec.store.parquet.TestEmptyParquet`.

**Comment by arina-ielchiieva** [comment]
> Ignores all schemas except last while writing empty parquet file
Please provide example...
> Not support empty schemas (e.g. create table .. as select * from empty.json, e.g. {})
What behavior will be in this case? Failure? No-op?

**Comment by oleg-zinovev** [comment]
> What behavior will be in this case? Failure? No-op?
Remained unchanged (No-op)

**Comment by oleg-zinovev** [comment]
> Please provide example...

I added a test TestParquetWriterEmptyFiles#testWriteEmptySchemaChange. As you can see, there is no "a" field in the written schema. 

Probably, it would be correct to write schema for all empty scans, but it will lead to writing "garbage" empty parquet files, if the scan with data is at the end of batch.

**Comment by arina-ielchiieva** [comment]
Is there a way to write combined schema in this case?

**Comment by oleg-zinovev** [comment]
> Please provide example...



> Is there a way to write combined schema in this case?

@arina-ielchiieva , thank you for your comment. 

I can try to make a combined scheme, but:
- What type of field should be written if the first BatchSchema contains field "A" with type "bigint", and the second - field "A" with type "varchar"? The last one wins?
- What about union vectors?

**Comment by arina-ielchiieva** [comment]
Good questions, you can investigate how now union types are handled. Regarding who wins, maybe you can look into `org.apache.drill.exec.physical.impl.union.UnionAllRecordBatch` to see how it Cretes combined schema using precedence rules.

**Comment by oleg-zinovev** [comment]
> Good questions, you can investigate how now union types are handled. Regarding who wins, maybe you can look into `org.apache.drill.exec.physical.impl.union.UnionAllRecordBatch` to see how it Cretes combined schema using precedence rules.

@arina-ielchiieva , thanks for your advice. 
I will try to make a combined scheme by the end of the week.

**Comment by oleg-zinovev** [comment]
@arina-ielchiieva 
Some time spent debugging the test showed that the last schema contains all fields. The field is added in ProjectRecordBatch#setupNewSchemaFromInput. 
In the original version of the test, field A was not added due to plan optimization - condition `1=0` was replaced by` limit 0`

I can still provide a solution with combining schema if required.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java [hunk 1]
```diff
@@ -486,11 +467,54 @@ public void abort() throws IOException {
...
+      parquetFileWriter = new ParquetFileWriter(conf, schema, path, ParquetFileWriter.Mode.OVERWRITE, blockSize, 0);
+    } else {
+      parquetFileWriter = new ParquetFileWriter(conf, schema, path, ParquetFileWriter.Mode.OVERWRITE);
+    }
+    storageStrategy.applyToFile(fs, path);
+    parquetFileWriter.start();
+  }
+
+  private void flushParquetFileWriter() throws IOException {
+    assert parquetFileWriter != null;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java]: Same here.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java [hunk 2]
```diff
@@ -486,11 +467,54 @@ public void abort() throws IOException {
...
   @Override
   public void cleanup() throws IOException {
-    flush();
+    flush(true);
 
     codecFactory.release();
   }
 
+  private void createParquetFileWriter() throws IOException {
+    assert parquetFileWriter == null;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java]: Please remove, there is no need for this check.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java [hunk 3]
```diff
@@ -310,27 +312,26 @@ public void checkForNewPartition(int index) {
...
+        flush(false);
         newSchema();
       }
     } catch (Exception e) {
       throw new DrillRuntimeException(e);
     }
   }
 
-  private void flush() throws IOException {
+  private void flush(final boolean cleanUp) throws IOException {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java]:
```suggestion
  private void flush(boolean cleanUp) throws IOException {
```

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java [hunk 4]
```diff
@@ -43,47 +49,99 @@ public void testWriteEmptyFile() throws Exception {
...
       // this specific value will force a flush just after the final row is written
       // this may cause the creation of a new "empty" parquet file
       test("ALTER SESSION SET `store.parquet.block-size` = 19926");
 
       final String query = "SELECT * FROM cp.`employee.json` LIMIT 100";
-      test("CREATE TABLE dfs.tmp.%s AS %s", outputFile, query);
+      test("CREATE TABLE dfs.tmp.%s AS %s", outputFileName, query);
+
+      // Make sure that only 1 parquet file was created
+      Assert.assertEquals(1, outputFile.list((dir, name) -> name.endsWith("parquet")).length);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: Please use static import.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java [hunk 5]
```diff
@@ -43,47 +49,99 @@ public void testWriteEmptyFile() throws Exception {
...
+            .build();
+
+    testBuilder()
+            .unOrdered()
+            .sqlQuery("select * from dfs.tmp.%s", outputFileName)
+            .schemaBaseLine(expectedSchema)
+            .go();
   }
 
   @Test // see DRILL-2408
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: Please remove `see DRILL-2408` references in the class, now the make no sense.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java [hunk 6]
```diff
@@ -43,47 +49,99 @@ public void testWriteEmptyFile() throws Exception {
...
-      testBuilder()
-        .unOrdered()
-        .sqlQuery("SELECT * FROM dfs.tmp.%s", outputFile)
-        .sqlBaselineQuery(query)
-        .go();
-    } finally {
-      runSQL("alter session set `planner.slice_target` = " + ExecConstants.SLICE_TARGET_DEFAULT);
-    }
+    // Make sure that only 1 parquet file was created
+    Assert.assertEquals(1, outputFile.list((dir, name) -> name.endsWith("parquet")).length);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: Static import.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java [hunk 7]
```diff
@@ -43,47 +49,99 @@ public void testWriteEmptyFile() throws Exception {
...
+  }
+
+  @Test
+  public void testEmptyFileSchema() throws Exception {
+    final String outputFileName = "testparquetwriteremptyfiles_testemptyfileschema";
+
+    test("CREATE TABLE dfs.tmp.%s AS SELECT * FROM cp.`employee.json` WHERE 1=0", outputFileName);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: Please replace select from JSON file with select from Parquet, we need to test that schema is created correctly from already known schema.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java [hunk 8]
```diff
@@ -122,6 +122,8 @@
   private PrimitiveTypeName logicalTypeForDecimals;
   private boolean usePrimitiveTypesForDecimals;
 
+  private boolean empty = true;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java]: Please add comment describing the purpose of this flag.

**Comment by arina-ielchiieva** [comment]
> @arina-ielchiieva
> Some time spent debugging the test showed that the last schema contains all fields. The field is added in ProjectRecordBatch#setupNewSchemaFromInput.
> In the original version of the test, field A was not added due to plan optimization - condition `1=0` was replaced by` limit 0`
> 
> I can still provide a solution with combining schema if required.

This case we don't need schema merge,

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java (hunk 7 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: If needed, I can also rewrite texts using RowSet.

**Comment by oleg-zinovev** [comment]
@arina-ielchiieva , thanks for review. Fixed.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java [hunk 9]
```diff
@@ -18,23 +18,33 @@
...
 public class TestParquetWriterEmptyFiles extends BaseTestQuery {
 
   @BeforeClass
   public static void initFs() throws Exception {
     updateTestCluster(3, null);
+    dirTestWatcher.copyResourceToRoot(Paths.get("schemachange"));
+    dirTestWatcher.copyResourceToRoot(Paths.get("parquet", "empty"));
   }
 
   @Test // see DRILL-2408
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: Please remove `see DRILL-2408`

Comment (oleg-zinovev) [reply exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: My bad. Done

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java [hunk 10]
```diff
@@ -122,6 +122,9 @@
   private PrimitiveTypeName logicalTypeForDecimals;
   private boolean usePrimitiveTypesForDecimals;
 
+  /** Whether no rows was written. */
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetRecordWriter.java]: Is used to ensure that empty Parquet file will be written if no rows were provided.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java [hunk 11]
```diff
@@ -18,72 +18,142 @@
...
+      .schemaBaseLine(expectedSchema)
+      .go();
+
+    // Make sure that only 1 parquet file was created
+    assertEquals(1, outputFile.list((dir, name) -> name.endsWith("parquet")).length);
   }
 
-  @Test // see DRILL-2408
+  @Test
+  public void testSimpleEmptyFileSchema() throws Exception {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: This test is redundant since `testComplexEmptyFileSchema` both cases.

Comment (arina-ielchiieva) [reply exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/writer/TestParquetWriterEmptyFiles.java]: Also we need to add test where we select from non-empty Parquet file but filter condition eliminates all rows, similar as you have for JSON.

**Comment by arina-ielchiieva** [comment]
@oleg-zinovev thanks for making the changes, a couple of minor comments are left...

**Comment by oleg-zinovev** [comment]
@arina-ielchiieva , thanks for review. Done

**Comment by arina-ielchiieva** [comment]
Ran all tests on the test cluster, all passed. LGTM, +1
@oleg-zinovev thanks for making the changes.
